odoo/addons/base/models/res_currency.py

```diff
@@ -120,20 +120,19 @@ def _check_company_currency_stays_active(self):
     def _get_rates(self, company, date):
         if not self.ids:
             return {}
-        currency_query = self.env['res.currency']._where_calc([
-            ('id', 'in', self.ids),
-        ], active_test=False)
+        currency_query = self._as_query(ordered=False)
         currency_id = self.env['res.currency']._field_to_sql(currency_query.table, 'id')
-        rate_query = self.env['res.currency.rate']._search([
+        Rate = self.env['res.currency.rate']
+        rate_query = Rate._search([
             ('name', '<=', date),
             ('company_id', 'in', (False, company.root_id.id)),
-            ('currency_id', '=', currency_id),
         ], order='company_id.id, name DESC', limit=1)
-        rate_fallback = self.env['res.currency.rate']._search([
+        rate_query.add_where(SQL("%s = %s", Rate._field_to_sql(rate_query.table, 'currency_id'), currency_id))
```

Comment:
What's the difference with `('currency_id', '=', currency_id)`?
AFAIK the query will be exactly the same (except that there is no possible domain optimization), but the syntax is harder to grasp IMO :thinking:

Reply:
I agree that the syntax is harder, however, you get the whole context of the query for which the SQL is created.
All optimizations are still applied, because when we saw SQL as a value, we could not apply any of them.

Now, we have a dedicated handler for custom domains - which are sometimes needed, but rarely and when it is needed, the caller has all arguments that were available in `_condition_to_sql`. It consolidates the conditions as domains albeit it has more arguments to keep all flexibility we had.